write files from helmit

edunham opened the pull request:
This drops the output of helmit into `chartsdir`, `charts` by default.

edunham committed:
json-e doesn't like installing for me without explicitly installing wheel first

edunham committed:
Helmit: write files

edunham committed:
Helmit: remove old prints

sciurus commented on infrastructure/k8s/bin/helmit.py:
```diff
@@ -75,13 +80,28 @@ def render_cronjob(project_name, secret_keys, deployment):
 parser = argparse.ArgumentParser()
 parser.add_argument("--service", help="Name of the service to render", default=None)
 args = parser.parse_args()
+chartsdir = "charts"
```
Instead of hardcoding this, add an argument via argparse that has this as the default.

sciurus commented on infrastructure/k8s/bin/helmit.py:
```diff
@@ -75,13 +80,28 @@ def render_cronjob(project_name, secret_keys, deployment):
+
+try:
+    os.mkdir(chartsdir)
+except:
```
Only `pass` on `FileExistsError`. Otherwise we risk ignoring errors like insufficient permissions to create the dir.

sciurus commented on infrastructure/k8s/requirements.txt:
```diff
@@ -1,2 +1,3 @@
+wheel==0.33.4
```
Why do you think we have a dependency on wheel? I don't see why we would.

edunham replied:
jsone fails to install in my python3 virtualenv without wheel installed, but succeeds with wheel.

edunham replied:
Er, previously failed. Tried again today and I can't reproduce the issue. Best guess is that I accidentally installed wheel outside the venv while troubleshooting and now it's available correctly on my system.

sciurus commented on infrastructure/k8s/bin/helmit.py:
```diff
@@ -60,9 +65,9 @@ def render_deployment(project_name, secret_keys, deployment):
-    print("---")
-    print(yaml.dump(jsone.render(template, context)))
-
+    # print("---")
```
Go ahead and delete this and the other commented-out lines.

sciurus commented on infrastructure/k8s/bin/helmit.py:
```diff
@@ -75,13 +80,28 @@ def render_cronjob(project_name, secret_keys, deployment):
+def write_file(template, context, prefix):
+    filename = chartsdir + "/taskcluster-" + prefix + "-" + p.split("/")[-1:][0]
+    try:
+        f = open(filename, "a+")
```
`a+` means that multiple runs would generate duplicate data.

Let's overwrite the files instead.

sciurus commented on infrastructure/k8s/bin/helmit.py:
```diff
@@ -75,13 +80,28 @@ def render_cronjob(project_name, secret_keys, deployment):
+    write_file(template, context, "cron")
+
+def write_file(template, context, prefix):
+    filename = chartsdir + "/taskcluster-" + prefix + "-" + p.split("/")[-1:][0]
```
I'm not sure exactly what you're going for here, since `p` is undefined.

Here's an idea I think would work:

For all charts, lets make the name be a combination of the `project_name` and the json-e file name (which matches the k8s type). For deployments, let's tack on the `proc_name`, and for cronjobs the `job_name` since there can be more than one of those things per service.

So e.g. for the purge-cache service we would have

* taskcluster-purge-cache-role.yaml
* taskcluster-purge-cache-role-binding.yaml
* taskcluster-purge-cache-service-account.yaml
*  taskcluster-purge-cache-deployment-web.yaml
* taskcluster-purge-cache-cron-job-expireCachePurges.yaml

edunham replied:
the rbacs split into "role", "role-binding", "service-account"

sciurus commented on infrastructure/k8s/bin/helmit.py:
```diff
@@ -75,13 +80,28 @@ def render_cronjob(project_name, secret_keys, deployment):
+        f.write(yaml.dump(jsone.render(template, context)))
+        f.close()
+    except: 
+        print("failed to write "+ filename)
```
Instead of building strings this way, use an f-string. See https://realpython.com/python-f-strings/#f-strings-a-new-and-improved-way-to-format-strings-in-python for an intro.

edunham committed:
Helmit: fixups

edunham committed:
Helmit: fix naming

$ ./bin/helmit.py --chartsdir charts
$ ls charts/
taskcluster-purge-cache-cron.yaml	   taskcluster-purge-cache-role.yaml		 taskcluster-references-deployment.yaml    taskcluster-references-secrets.yaml
taskcluster-purge-cache-deployment.yaml    taskcluster-purge-cache-secrets.yaml		 taskcluster-references-role-binding.yaml  taskcluster-references-service-account.yaml
taskcluster-purge-cache-role-binding.yaml  taskcluster-purge-cache-service-account.yaml  taskcluster-references-role.yaml

edunham committed:
Revert "json-e doesn't like installing for me without explicitly installing wheel first"

This reverts commit b1cb7e7f3391e95e02059878ee15cb00a397f104.

sciurus reviewed (changes requested):
Commented on slack

edunham committed:
Helmit: Names compatible with multiple crons

$ ./bin/helmit.py
$ ls charts/
taskcluster-purge-cache-deployment-web.yaml	     taskcluster-references-deployment-web.yaml
taskcluster-purge-cache-expireCachePurges-cron.yaml  taskcluster-references-role-binding.yaml
taskcluster-purge-cache-role-binding.yaml	     taskcluster-references-role.yaml
taskcluster-purge-cache-role.yaml		     taskcluster-references-secrets.yaml
taskcluster-purge-cache-secrets.yaml		     taskcluster-references-service-account.yaml
taskcluster-purge-cache-service-account.yaml

edunham committed:
Helmit: consistent naming
